drivers/rilmodem/radio-settings.c

```diff
@@ -87,16 +149,339 @@ void ril_set_rat_mode(struct ofono_radio_settings *rs,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL) {
+		ofono_error("%s: parse error", __func__);
+		return;
+	}
+
+	if (sd->pending_msgs != 0)
+		return;
+
+	ofono_info("Switching radio caps between slots - FINISH");
+
+	set_preferred_network(rd1, sd->cbd, sd->mode_to_switch);
+
+	/*
+	 * If the complementary slot does not support anymore its current
+	 * technology, we change it to the best possible among available ones.
+	 */
+	if ((rd2->rat_mode & rd2->available_rats) == 0) {
+
+		struct cb_data *cbd =
+			cb_data_new(set_preferred_cb, rd2, rd2->radio_settings);
+
+		set_preferred_network(rd2, cbd,
+				get_best_available_tech(rd2->available_rats));
+	}
+
+	rd1->switch_d = NULL;
+	rd2->switch_d = NULL;
+	g_free(sd);
+	g_free(caps);
+}
+
+static void radio_caps_event(struct ril_msg *message, gpointer user_data)
+{
+	struct radio_data *rd = user_data;
+	struct switch_data *sd = rd->switch_d;
+	struct radio_data *rd1 = sd->rd_1;
+	struct radio_data *rd2 = sd->rd_2;
+	struct reply_radio_capability *caps;
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL) {
+		ofono_error("%s: parse error", __func__);
+		return;
+	}
+
+	if (sd == NULL)
+		return;
+
+	/*
+	 * Update rats. They come also in the replies to SET_RADIO_CAPABILITY,
+	 * but those seem to be unreliable, at least for midori.
+	 */
+	rd->ril_rats = caps->rat;
+	rd->available_rats = set_rat_from_ril_rat(caps->rat);
+
+	strcpy(rd->modem_uuid, caps->modem_uuid);
+
+	sd->pending_msgs--;
+
+	if (sd->pending_msgs != 0)
+		return;
+
+	DBG("Sending requests for FINISH phase");
+
+	send_set_radio_cap(rd1, g_session, RIL_RC_PHASE_FINISH,
+				rd1->ril_rats, rd1->modem_uuid,
+				RIL_RC_STATUS_SUCCESS, switch_finish_cb);
+	send_set_radio_cap(rd2, g_session, RIL_RC_PHASE_FINISH,
+				rd2->ril_rats, rd2->modem_uuid,
+				RIL_RC_STATUS_SUCCESS, switch_finish_cb);
+	sd->pending_msgs = 2;
+
+	g_free(caps);
+}
+
+/*
+ * This function is just for completeness, as we actually need to wait for the
+ * unsolocited events to continue the capabilities switch.
+ */
+static void switch_apply_cb(struct ril_msg *message, gpointer user_data)
+{
+	struct radio_data *rd = user_data;
+	struct reply_radio_capability *caps;
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: error %s", __func__,
+				ril_error_to_string(message->error));
+		return;
+	}
+
+	caps = g_ril_reply_parse_get_radio_capability(rd->ril, message);
+	if (caps == NULL)
+		ofono_error("%s: parse error", __func__);
+
+	g_free(caps);
+}
+
+static void switch_start_cb(struct ril_msg *message, gpointer user_data)
+{
+	struct radio_data *rd = user_data;
+	struct switch_data *sd = rd->switch_d;
+	struct radio_data *rd1 = sd->rd_1;
+	struct radio_data *rd2 = sd->rd_2;
+	struct reply_radio_capability *caps;
+
+	sd->pending_msgs--;
+
+	if (message->error != RIL_E_SUCCESS) {
+		ofono_error("%s: error %s", __func__,
```

Comment:
If the callback for the second modem instance fails here, there's no cleanup of rd->switch_d, so the next time the ril_set_rat_mode is called ( when a user tries to set the preferred tech ), it'll fail because rd->switch_d is non NULL.